### DISCUSSION TITLE
Polymorphic Relationships

### DISCUSSION THREAD

**Discussion by woren951** [adonisjs/core#4661] [Ideas]
Hi, I'm new to adonisjs, moved here after using Laravel for a few years, read the documentation and really liked everything. But didn't find support for [polymorphic relationships](https://laravel.com/docs/11.x/eloquent-relationships#polymorphic-relationships), is there any future support for these?

**Comment by holoyan** [comment]
any news one this???

**Reply by MohammedElattar** [reply]
https://github.com/adonisjs/lucid/issues/86#issuecomment-2784748661

**Comment by syl-p** [comment]
Hello @woren951  and @holoyan,
I think, I find a way to mimic the polymorphic relationships (like ruby on rails or laravel). I my exemple I want to add Tag to Post, Event, Video or whatever.

Migration:
```
export default class extends BaseSchema {
  protected tableName = 'tags'

  async up() {
    this.schema.createTable(this.tableName, (table) => {
      table.increments('id')
      table.string('label')
      table.integer('taggable_id')
      table.string('taggable_type')
      table.timestamp('created_at')
      table.timestamp('updated_at')
    })
  }

  async down() {
    this.schema.dropTable(this.tableName)
  }
}
```

Tag Model:
```
export default class Tag extends BaseModel {
  @column({ isPrimary: true })
  declare id: number

  @column.dateTime({ autoCreate: true })
  declare createdAt: DateTime

  @column.dateTime({ autoCreate: true, autoUpdate: true })
  declare updatedAt: DateTime

  @column()
  declare label: string

  @column()
  declare taggableType: string

  @column()
  declare taggableId: number

  @computed()
  get relatedEntity() {
    const ModelClass = this.getModelClass(this.taggableType)
    return ModelClass[this.taggableType]
  }

  private getModelClass(type: string) {
    const models = {
      Post: () => import('#models/post'),
      Event: () => import('#models/event'),
    }

    return models[type]?.().then((module) => module.default)
  }
}

```

Post model
```
  @hasMany(() => Tag, {
    foreignKey: 'taggableId',
    onQuery: (query) => query.where('taggable_type', 'Post'),
  })
  declare tags: HasMany<typeof Tag>
```

Don't forget to set taggable_type to your tag when you are creating it:

```
await post.related('tags').create({
  label: 'Super article !',
  taggableType: 'Post',
})
```

```
await event.related('tags').create({
  label: 'Super article !',
  taggableType: 'Event',
})
```

does this answer your question ?

**Reply by holoyan** [reply]
Hi @syl-p , thanks for sharing it but this solution works only one side  right? - `post.tags` or `video.tags` and so on. I'm looking for  a reverse direction, exmaple

```typescript

const tag = await Tag.query().preload('morphs').first()

// and here loop over tag.morphs, where morphs can be post,video model or something esle


```

**Reply by syl-p** [reply]
:thinking: You are right ! Could we hack the relation by construct manually the query with a WHERE taggable_type = "MorphModel" ? 

For the moment, I don't try to made the reverse direction.

**Reply by holoyan** [reply]
not sure how we can do that, because for reverse relation you can't specify [target](https://laravel.com/docs/11.x/eloquent-relationships#one-to-many-polymorphic-relations) model, 
```typescript


  @morpthTo()
  declare taggable: MorphTo


```

**Comment by MohammedElattar** [comment]
https://github.com/adonisjs/lucid/issues/86#issuecomment-2784748661
